lib/Monads/NonDetMonad.thy

```diff
@@ -722,6 +722,38 @@ where
 notation (output)
   whileLoopE  ("(whileLoopE (_)//  (_))" [1000, 1000] 1000)
 
+
+section "More Monadic Combinators"
+
+definition
+  whileM :: "('s, bool) nondet_monad \<Rightarrow> ('s, 'a) nondet_monad \<Rightarrow> ('s, unit) nondet_monad" where
+  "whileM C B \<equiv> do
+    c \<leftarrow> C;
+    whileLoop (\<lambda>c s. c) (\<lambda>_. do B; C od) c;
+    return ()
+  od"
+
+definition
```

Comment:
this one could use a comment for future people to understand

Reply:
I think I wrote that one. What comment would you like? It's `ifM`..

Reply:
(perfectly happy to add something, just not sure which bit is unclear, since I wrote it :-))

Reply:
I think the section title / comment for combinators with conditions that have side effects will take care of this, as it then becomes clear what's being done here.